clang/include/clang/Sema/BoundsWideningAnalysis.h

```diff
@@ -15,55 +15,442 @@
+    void DumpWidenedBounds(FunctionDecl *FD);
+
+    // Get the Out set for the statement. This set represents the bounds
+    // widened after the statement.
+    // Note: This method can be called from outside this class to get the
+    // widened bounds after a statement.
+    // @param[in] B is the current block.
+    // @param[in] CurrStmt is the current statement.
+    BoundsMapTy GetStmtOut(const CFGBlock *B, const Stmt *CurrStmt) const;
+
+    // Get the In set for the statement. This set represents the bounds
+    // widened before the statement.
+    // Note: This method can be called from outside this class to get the
+    // widened bounds before a statement.
+    // @param[in] B is the current block.
+    // @param[in] CurrStmt is the current statement.
+    BoundsMapTy GetStmtIn(const CFGBlock *B, const Stmt *CurrStmt) const;
+
+  private:
+    // Compute Gen and Kill sets for the block and statements in the block.
+    // @param[in] EB is the current ElevatedCFGBlock.
+    // @param[in] NestedStmts is a set of top-level statements that are
+    // nested in another top-level statement.
+    void ComputeGenKillSets(ElevatedCFGBlock *EB, StmtSetTy NestedStmts);
+
+    // Compute the StmtGen and StmtKill sets for a statement in a block.
+    // @param[in] EB is the current ElevatedCFGBlock.
+    // @param[in] CurrStmt is the current statement.
+    // @param[in] NestedStmts is a set of top-level statements that are
+    // nested in another top-level statement.
+    void ComputeStmtGenKillSets(ElevatedCFGBlock *EB, const Stmt *CurrStmt,
+                                StmtSetTy NestedStmts);
+
+    // Compute the union of Gen and Kill sets of all statements up to (and
+    // including) the current statement in the block.
+    // @param[in] EB is the current ElevatedCFGBlock.
+    // @param[in] CurrStmt is the current statement.
+    // @param[in] PrevStmt is the previous statement of CurrStmt in the linear
+    // ordering of statements in the block.
+    void ComputeUnionGenKillSets(ElevatedCFGBlock *EB, const Stmt *CurrStmt,
+                                 const Stmt *PrevStmt);
+
+    // Compute the Gen and Kill sets for the block.
+    // @param[in] EB is the current ElevatedCFGBlock.
+    void ComputeBlockGenKillSets(ElevatedCFGBlock *EB);
+
+    // Compute the In set for the block.
+    // @param[in] EB is the current ElevatedCFGBlock.
+    void ComputeInSet(ElevatedCFGBlock *EB);
+
+    // Compute the Out set for the block.
+    // @param[in] EB is the current ElevatedCFGBlock.
+    // @param[out] EB is added to WorkList if the Out set of EB changes.
+    void ComputeOutSet(ElevatedCFGBlock *EB, WorkListTy &WorkList);
+
+    // Initialize the In and Out sets for the block.
+    // @param[in] FD is the current function.
+    // @param[in] EB is the current ElevatedCFGBlock.
+    void InitBlockInOutSets(FunctionDecl *FD, ElevatedCFGBlock *EB);
+
+    // Prune the Out set of the pred block according to various conditions.
```

Comment:
Here, it will be great to have an explanation about why `PruneOutSet` is required because creating this function is an implementation decision. As I understand, the implementation of  the intersection operation as defined in the `BoundsWideningAnalysis` is done in two steps:
1) the function `PruneOutSet` pre-processes the `BoundsMaps` that we wish to intersect, and
2) the regular intersection operation is performed.